elastic/kibana
# Apply back pressure in Task Manager whenever Elasticsearch responds with a 429

Resolves #65553.

In this PR, I'm implementing the proposal discussed [here](https://github.com/elastic/kibana/issues/65553#issuecomment-674076858).

This PR represents a feature branch where pieces of work are getting merged:
- [x] Make task manager maxWorkers and pollInterval observables #75293
- [x] [to master branch] Add support for decorating and handling 429 errors with the saved objects client #75664
- [x] Make task manager store emit error events #75679
- [x] Temporarily apply back pressure to maxWorkers and pollInterval when 429 errors occur #77096

### How to test this locally?

1. Start up Elasticsearch with a setting to persist the data `yarn es snapshot --ssl -E path.data=/my/path`
2. Start up Kibana (`yarn start --ssl`)
3. Wait for Kibana to be ready
4. Stop Kibana
5. Stop Elasticsearch
6. Start Elasticsearch with a thread pool limit `yarn es snapshot --ssl -E path.data=/my/path -E thread_pool.write.queue_size=3`
7. Somewhere in Kibana, register an empty task definition in task manager on plugin setup
8. Somewhere in Kibana, simultaneously schedule 200-1000 tasks using the type created above on plugin start
9. Start Kibana (`yarn start --ssl`)
10. Warning logs should appear shortly after Kibana is ready

## Note for docs

This is worth documenting somewhere that Task Manager will reduce max workers and increase poll interval whenever Elasticsearch responds with 429 errors on task manager CRUD operations.
